### PR TITLE
Markdownlint - Add custom rules

### PR DESCRIPTION
**Pull request by thatblindgeye**
## Because

The default rules for Markdownlint were missing some things that our style guide mentions

## This PR

- Adds various custom rules, including a modified heading structure rule
  - TOP001 will catch inaccessible link text that commonly includes things such as "this video", "click here", etc. It won't say what to update the link text to, because the context is super important, but flagging it is a huge time saver
  - TOP002 does what it says: complains if we try using inline code in headings
  - TOP003 also does what it says: for sections like "lesson overview" where we are using default text over and over, the rule complains when that content doesn't match up. It'll catch if the post heading verbiage isn't there/is wrong, if there is no unordered list following that verbiage, and if there is extra content after or before the unordered list.
  - TOP004 is just a slightly tweaked version of the default linter's required-headings rule. Where the default rule only had a general wild card `*`, ours allows something like `#### *` to say "any heading is allowed here as long as it's a level 4 heading"

## Issue

Closes #XXXXX

## Additional Information

For the link text rule, it may catch some "false positives", such as something like `[This blog post on flex-grow from CoolBlogSite](...url)`. Originally I went with an arbitrary "throw an error if there's less than 5 words before and/or after instances of `this`", but like I said it's arbitrary and may not be any better. I figure for these cases, we should still update the link text to pull "this" out, so that we end up with `this blog post on [flex-grow from CoolBlogSite](...url)` or something.

## Pull Request Requirements

-   [ ] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [ ] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [ ] The `Because` section summarizes the reason for this PR
-   [ ] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

File: markdownlint/TOP001_descriptiveLinkText.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    );
+    const linkOpenTokens = linesWithLinks
+      .map((line) =>
+        line.children?.filter((child) => child.type === "link_open")
+      )
+      .flat();
+
+    linkOpenTokens.forEach((token) => {
+      const { line } = token;
+      const isInvalid = /\[(.+\s*)?(this|here)(\s*.+)?\]/i.test(line);
```
Comment (thatblindgeye) [inline markdownlint/TOP001_descriptiveLinkText.js:17]: Link to Regexr to show valid and invalid examples

File: markdownlint/TOP004_lessonHeadings.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    let hasError = false;
+    let anyHeadings = false;
+    const getExpected = () => requiredHeadings[i++] || "[None]";
+    const handleCase = (str) => (matchCase ? str : str.toLowerCase());
+    forEachHeading(params, (heading, content) => {
+      if (!hasError) {
+        anyHeadings = true;
+        const actual = levels[heading.tag] + " " + content;
+        const expected = getExpected();
+        if (/^#*\s?\*$/.test(expected)) {
```
Comment (thatblindgeye) [inline markdownlint/TOP004_lessonHeadings.js:101]: Same as above here, also store as a new Regex

File: .markdownlint-cli2.jsonc [hunk 3]
```diff
@@ -79,5 +79,11 @@
     "MD050": {
       "style": "asterisk"
     }
-  }
+  },
+  "customRules": [
```
Comment (thatblindgeye) [inline .markdownlint-cli2.jsonc:83]: Add some test markdown files that can be checked when custom rules are added/updated

**Comment by thatblindgeye** [comment]
@wise-king-sullyman So bit of a hiccup with the testing MD files, at least in this PR. I tested each file locally and everything that should be flagged did get flagged, and anything that should NOT be flagged wasn't. There's some limitation with some of these rules (custom or default) when the GH action runs, though. I can try opening a separate PR for each of these test files and close them to show they work as intended.

**Review by wise-king-sullyman — APPROVED** [review]
I'm good with however you want to proceed. IMO followup PRs just doing whitespace changes or something to the test files that aren't getting flagged properly would be fine.

**Pull request merged**